csrc/multidevice/utils.cpp

```diff
@@ -13,22 +13,93 @@
-  std::vector<IterDomain*> sharded_ids;
+std::unordered_set<IterDomain*> getShardedIterDomains(TensorView* tv) {
+  std::unordered_set<IterDomain*> sharded_ids;
   std::copy_if(
       tv->getLeafDomain().begin(),
       tv->getLeafDomain().end(),
-      std::back_inserter(sharded_ids),
+      std::inserter(sharded_ids, sharded_ids.begin()),
       [](auto id) { return id->isDeviceDim(); });
   return sharded_ids;
 }
+
+// Returns whether a IterDomain in a TensorView can be read/written
+// contiguously. when a IterDomain's sharding is flipped. This occurs when the
+// IterDomain being unsharded/sharded is the outermost in the allocation domain.
+// e.g. [DIDx(i0), r1, i2, i3]
+// i0 => true, outermost axis
+// i1 => true, reduction axis are not allocated
+// i2 => true, reduction and device axis are not allocated
+// i3 => false, i2 is allocated.
+bool isContiguousShard(TensorView* tv, IterDomain* changed_id) {
+  bool not_allocated = true;
+  for (auto id : tv->getLeafDomain()) {
+    if (id == changed_id) {
+      return not_allocated;
+    }
+    not_allocated = not_allocated && (id->isDeviceDim() || id->isReduction());
+  }
+  NVF_ERROR(
+      false,
+      "Id",
+      changed_id->toString(),
+      " is not in TensorView ",
+      tv->toString());
+  return false;
+}
+
+// For a resharding expression, either a set or reduce, returns root IDs
+// that change sharding.
+// (1) sharded root IterDomains that are added by the expression
+// i.e. sharded IterDomains that are present in the output, but not the input.
+// (2) sharded root IterDomains that are removed by the expression
+// i.e. sharded IterDomains that are present in the input, but not the output.
+// TODO: Analyze leaf domain for unsharded/sharded IDs and return their
+// parent root IDs.
+std::pair<std::vector<IterDomain*>, std::vector<IterDomain*>>
+allocationShardings(Expr* expr) {
+  NVF_ERROR(
+      expr->outputs().size() == 1,
+      "Resharding operations can only have one output");
+  NVF_ERROR(
+      expr->inputs().size() == 1,
+      "Resharding operations can have only one input");
+  auto output = expr->outputs().at(0)->as<TensorView>();
+  auto input = expr->inputs().at(0)->as<TensorView>();
+
+  auto sharded_ids_input = getShardedIterDomains(input);
+  auto sharded_ids_output = getShardedIterDomains(output);
+  std::vector<IterDomain*> shard_additions;
+  std::vector<IterDomain*> shard_deletions;
+  auto rootmap = PairwiseRootDomainMap(input, output);
+
+  const auto c2p_map = rootmap.mapConsumerToProducer(&sharded_ids_output);
+  for (auto [id1, id2] : c2p_map) {
+    if (!id1->sameAs(id2)) {
+      shard_additions.push_back(id1);
+    }
+  }
+
+  const auto p2c_map = rootmap.mapProducerToConsumer(&sharded_ids_input);
+  for (auto [id1, id2] : p2c_map) {
+    // Ignore sharded reductions i.e.
+    // DIDx(i0) -> r(i0) or DIDx(i0)->r(DIDx(i0))
+    // since they don't affect allocation.
+    if (!id1->sameAs(id2) && !id2->isReduction()) {
```

Comment:
We only have `DIDx` for now, but what if `id1`'s parallel type is `DIDx` and the reduction `id2` is `DIDy`?